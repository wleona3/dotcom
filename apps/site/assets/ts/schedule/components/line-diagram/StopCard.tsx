import React, { ReactElement, useContext } from "react";
import { LineDiagramStop, RouteStop } from "../__schedule";
import { isACommuterRailRoute, isAGreenLineRoute } from "../../../models/route";
import { LiveData } from "./__line-diagram";
import { isMergeStop, diagramWidth } from "./line-diagram-helpers";
import StopConnections from "./StopConnections";
import StopPredictions from "./StopPredictions";
import { hasPredictionTime } from "../../../models/prediction";
import { alertIcon } from "../../../helpers/icon";
import {
  isHighSeverityOrHighPriority,
  isDiversion,
  isCurrentAlert
} from "../../../models/alert";
import { Alert, Route } from "../../../__v3api";
import MatchHighlight from "../../../components/MatchHighlight";
import StopFeatures from "./StopFeatures";
import { StopRefContext } from "./LineDiagramWithStops";
import { effectNameForAlert } from "../../../components/Alerts";
import GlxOpen from "../../../components/GlxOpen";

interface StopCardProps {
  stop: LineDiagramStop;
  onClick: (stop: RouteStop) => void;
  liveData?: LiveData;
  searchQuery?: string;
}

const hasBranchLabel = (stop: LineDiagramStop): boolean =>
  stop.route_stop["is_terminus?"] &&
  !!stop.route_stop.branch &&
  !!stop.route_stop.route;

const lineName = ({ name, route: routeStopRoute }: RouteStop): string => {
  const route = routeStopRoute as Route;
  const title = isAGreenLineRoute(route)
    ? `Green Line ${route.id.split("-")[1]}`
    : name;
  const lineOrBranch = isACommuterRailRoute(route) ? "Line" : "Branch";
  return `${title} ${lineOrBranch}`;
};

// const MaybeAlert = (alerts: Alert[]): JSX.Element | null => {
//   const highPriorityAlerts = alerts.filter(isHighSeverityOrHighPriority);
//   if (highPriorityAlerts.length === 0) return null;
//   return (
//     <>
//       {alertIcon("c-svg__icon-alerts-triangle")}
//       <span className="sr-only">Service alert or delay</span>
//       &nbsp;
//     </>
//   );
// };

const StopCard = (props: StopCardProps): ReactElement<HTMLElement> => {
  const { stop, onClick, liveData, searchQuery } = props;
  const {
    stop_data: stopData,
    route_stop: routeStop,
    alerts: stopAlerts,
    route_stop: { "is_beginning?": isOrigin, "is_terminus?": isTerminus }
  } = stop;

  const isDestination = !isOrigin && isTerminus;
  const width = isMergeStop(stop)
    ? diagramWidth(1)
    : diagramWidth(stopData.length);
  const refs = useContext(StopRefContext)[0];

  // const diversionAlert = stopAlerts.find(
  //   alert => isDiversion(alert) && isCurrentAlert(alert)
  // );
  const diversionAlert = {
    "id": "2",
    "severity": 7,
    "effect": "shuttle",
    "lifecycle": "ongoing",
    "active_period": [
      ["2020-09-01 12:00", "2023-01-01 01:00"]
    ]
  } as Alert;

  const hasLivePredictions =
    liveData &&
    liveData.headsigns.length &&
    liveData.headsigns.some(hasPredictionTime);
  const showPrediction = hasLivePredictions && !isDestination;
  const showDiversion =
    diversionAlert && !(hasLivePredictions && isDestination);

  return (
    <li
      className="m-schedule-diagram__stop"
      style={{ paddingLeft: searchQuery ? "0.5rem" : `${diagramWidth(2)}px` }}
    >
      <section className="m-schedule-diagram__content">
        <GlxOpen pageType="line-diagram" stopId={routeStop.id} />
        {hasBranchLabel(stop) && (
          <div className="u-bold u-small-caps">{lineName(routeStop)}</div>
        )}
        <header
          className="m-schedule-diagram__stop-heading"
          ref={el => refs.set(routeStop.id, el)}
        >
          <h4 className="m-schedule-diagram__stop-link notranslate">
            <a href={`/stops/${routeStop.id}`}>
              {/* {MaybeAlert([{
                "id": "2",
                "severity": 7,
                "effect": "detour",
                "lifecycle": "ongoing",
                // "active_period": [
                //   ["2020-09-01 12:00", "2023-01-01 01:00"]
                // ]
              } as Alert])} */}
              <MatchHighlight text={routeStop.name} matchQuery={searchQuery} />
            </a>
          </h4>
          {StopFeatures(routeStop)}
        </header>

        <div className="m-schedule-diagram__stop-details">
          {StopConnections(routeStop.connections)}
          {/* {showPrediction ? (
            <StopPredictions
              headsigns={liveData!.headsigns}
              isCommuterRail={
                !!routeStop.route && isACommuterRailRoute(routeStop.route)
              }
            />
          ) : (
            showDiversion && (
              <div
                key={diversionAlert!.id}
                className="m-schedule-diagram__alert"
              >
                {effectNameForAlert(diversionAlert!)}
              </div>
            )
          )} */}
          <div
            key={diversionAlert!.id}
            className="m-schedule-diagram__alert"
          >
            {effectNameForAlert(diversionAlert!)}
          </div>
        </div>

        {/* <footer className="m-schedule-diagram__footer">
          <button
            className="btn btn-link"
            type="button"
            onClick={() => onClick(routeStop)}
          >
            View schedule
          </button>
        </footer> */}
      </section>
    </li>
  );
};

export default StopCard;
