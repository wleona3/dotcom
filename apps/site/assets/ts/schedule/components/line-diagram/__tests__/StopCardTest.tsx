import React from "react";
import { cloneDeep, merge } from "lodash";
import {
  RouteType,
  HeadsignWithCrowding,
  Schedule,
  Prediction
} from "../../../../__v3api";
import { LineDiagramStop } from "../../__schedule";
import simpleLineDiagram from "./lineDiagramData/simple.json"; // not a full line diagram
import outwardLineDiagram from "./lineDiagramData/outward.json"; // not a full line diagram
import StopCard from "../StopCard";
import { TripPrediction } from "../../__trips";
import { mockedRenderWithStopPositionContext } from "../../../../__tests__/test-helpers";

const lineDiagram = (simpleLineDiagram as unknown) as LineDiagramStop[];
let lineDiagramBranchingOut = (outwardLineDiagram as unknown) as LineDiagramStop[];

const route = {
  type: 3 as RouteType,
  name: "route 1",
  long_name: "route 1 long name",
  color: "F00B42",
  id: "route-1",
  direction_names: {
    0: "Outbound",
    1: "Inbound"
  },
  direction_destinations: {
    0: "Begin",
    1: "End"
  },
  description: "key_bus_route",
  "custom_route?": false,
  header: "",
  alerts: []
};

lineDiagram.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

lineDiagramBranchingOut.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

let lineDiagramBranchingIn = cloneDeep(lineDiagramBranchingOut).reverse();
const CRroute = merge(cloneDeep(route), { type: 2 as RouteType });
lineDiagramBranchingIn.forEach(({ route_stop }) => {
  route_stop.route = CRroute;
  if (route_stop["is_terminus?"]) {
    route_stop["is_beginning?"] = !route_stop["is_beginning?"];
  }
});

const handleStopClick = () => {};
const liveData = { headsigns: [], vehicles: [] };

describe("StopCard", () => {
  let asFragment: () => DocumentFragment, container: HTMLElement;
  beforeEach(() => {
    ({ asFragment, container } = mockedRenderWithStopPositionContext(
      <StopCard
        stop={lineDiagram[0]}
        onClick={handleStopClick}
        liveData={liveData}
      />,
      lineDiagram
    ));
  });

  it("renders and matches snapshot", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("includes a button to open Schedule Finder on each stop", () => {
    expect(
      container.querySelector(".m-schedule-diagram__footer > button")
    ).toBeTruthy();
    expect(
      container.querySelector(".m-schedule-diagram__footer > button")
        ?.textContent
    ).toContain("View schedule");
  });

  it("has a tooltip for a transit connection", () => {
    const stopConnections = container.querySelectorAll<HTMLElement>(
      ".m-schedule-diagram__connections a"
    );
    stopConnections.forEach(connectionLink => {
      expect(connectionLink.dataset.toggle).toEqual("tooltip");
      expect(connectionLink.dataset.originalTitle).toContain("Route");
    });
  });

  it("indicates detours, stop closures, etc", () => {
    expect(container.querySelector(".m-schedule-diagram__alert")).toBeTruthy();
    expect(
      container.querySelector(".m-schedule-diagram__alert")?.textContent
    ).toContain("Detour");
  });
});

const predictionHeadsign: HeadsignWithCrowding = {
  name: "Somewhere",
  time_data_with_crowding_list: [
    {
      time_data: {
        delay: 0,
        scheduled_time: ["4:30", " ", "PM"],
        prediction: {
          time: ["14", " ", "min"],
          status: null,
          track: null
        } as Prediction
      },
      crowding: null,
      predicted_schedule: {
        schedule: {} as Schedule,
        prediction: {} as TripPrediction
      }
    }
  ],
  train_number: null
};
const liveDataWithPrediction = {
  headsigns: [predictionHeadsign],
  vehicles: []
};
it("indicates predictions if available", () => {
  const { asFragment, container } = mockedRenderWithStopPositionContext(
    <StopCard
      stop={lineDiagram[2]}
      onClick={handleStopClick}
      liveData={liveDataWithPrediction}
    />,
    lineDiagram
  );

  it("renders and matches snapshot", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  const predictions = container.querySelector(
    ".m-schedule-diagram__predictions"
  );
  expect(predictions).toBeTruthy();
  expect(predictions?.textContent).toContain("Somewhere");
  expect(
    predictions!.querySelector(".m-schedule-diagram__prediction-time")
      ?.textContent
  ).toContain("14");
  expect(
    predictions!.querySelector(".m-schedule-diagram__prediction-time")
      ?.textContent
  ).toContain("min");
});

it.each`
  index | expectedAlerts
  ${0}  | ${0}
  ${1}  | ${1}
  ${2}  | ${1}
  ${3}  | ${0}
`(
  "shows $expectedAlerts high priority or high severity alerts for stop $index",
  ({ index, expectedAlerts }) => {
    const { container } = mockedRenderWithStopPositionContext(
      <StopCard
        stop={lineDiagram[index]}
        onClick={handleStopClick}
        liveData={liveData}
      />,
      lineDiagram
    );

    const alerts = container.querySelectorAll(
      ".m-schedule-diagram__stop-link .c-svg__icon-alerts-triangle"
    );
    expect(alerts.length).toEqual(expectedAlerts);
  }
);

it.each`
  index | expectedNames                      | expectedFeatures
  ${0}  | ${[]}                              | ${["Parking"]}
  ${1}  | ${["Orange Line", "Green Line C"]} | ${[]}
  ${2}  | ${["Route 62", "Route 67"]}        | ${["Accessible"]}
  ${3}  | ${["Atlantis"]}                    | ${["Parking", "Accessible"]}
`(
  "has appropriate tooltip content for stop $index",
  ({ index, expectedNames, expectedFeatures }) => {
    const { container } = mockedRenderWithStopPositionContext(
      <StopCard
        stop={lineDiagram[index]}
        onClick={handleStopClick}
        liveData={liveData}
      />,
      lineDiagram
    );

    const connections = container.querySelector(
      ".m-schedule-diagram__connections"
    );
    const names = Array.from(connections!.querySelectorAll("a")).map(
      c => c.dataset.originalTitle
    );
    expect(names).toEqual(expectedNames);

    const features = container.querySelector(".m-schedule-diagram__features");

    const featureNames = Array.from(
      features!.querySelectorAll<HTMLElement>("span[data-toggle='tooltip']")
    ).map(c => c.dataset.originalTitle);
    expect(featureNames).toEqual(expectedFeatures);
  }
);
