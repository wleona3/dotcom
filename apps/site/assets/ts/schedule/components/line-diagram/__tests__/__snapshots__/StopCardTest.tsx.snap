// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopCard renders and matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="m-schedule-diagram__stop"
    style="padding-left: 36px;"
  >
    <section
      class="m-schedule-diagram__content"
    >
      <header
        class="m-schedule-diagram__stop-heading"
      >
        <h4
          class="m-schedule-diagram__stop-link notranslate"
        >
          <a
            href="/stops/line-origin"
          >
            <span>
              Origin
            </span>
          </a>
        </h4>
        <div
          class="m-schedule-diagram__features"
        >
          <span
            data-animation="true"
            data-html="false"
            data-original-title="Parking"
            data-placement="bottom"
            data-selector="true"
            data-toggle="tooltip"
            data-trigger="hover focus"
          >
            <i
              aria-hidden="false"
              class="fa fa-square-parking notranslate m-schedule-diagram__feature-icon"
            />
          </span>
          <span
            data-animation="true"
            data-html="false"
            data-original-title="Accessible"
            data-placement="bottom"
            data-selector="true"
            data-toggle="tooltip"
            data-trigger="hover focus"
          >
            <span
              aria-hidden="false"
              class="notranslate c-svg__icon-acessible-default m-schedule-diagram__feature-icon"
            >
              SVG
            </span>
          </span>
        </div>
      </header>
      <div
        class="m-schedule-diagram__stop-details"
      >
        <div
          class="m-schedule-diagram__connections"
        />
        <div
          class="m-schedule-diagram__alert"
        >
          Detour
        </div>
      </div>
      <footer
        class="m-schedule-diagram__footer"
      >
        <button
          class="btn btn-link"
          type="button"
        >
          View schedule
        </button>
      </footer>
    </section>
  </li>
</DocumentFragment>
`;
